### ISSUE TITLE
customize loss function for QAOA

### ISSUE THREAD

**Issue by shangtai** [qiboteam/qibo#635]
I am trying to customize loss function for QAOA. I have changed the files in variational.py and hamiltonians.py in customize_qaoa_loss branch. 

  In the minimization function of QAOA, I have introduced another input argument known as loss 
```def 
minimize(
          self,
          initial_p,
          initial_state=None,
          method="Powell",
          loss=None,
          jac=None,
          hess=None,
          hessp=None,
          bounds=None,
          constraints=(),
          tol=None,
          callback=None,
          options=None,
          compile=False,
          processes=None,
      )
```

and in _loss, I tried to introduce if-else statements to consider the alternatives.

```
def _loss(params, qaoa, hamiltonian, state):
            if state is not None:
                state = hamiltonian.backend.cast(state, copy=True)
            qaoa.set_parameters(params)
            state = qaoa(state)
            if loss is None:  # How do I make this a default mode usually it is something like <function QAOA.minimize.<locals>.<lambda> at 0x14777fdc0>
                return hamiltonian.expectation(state)
            elif loss == "cvar":
                return hamiltonian.cvar(state)
            elif loss == "gibbs":
                return hamiltonian.gibbs(state)
```

The loss in _loss is something like <function QAOA.minimize.<locals>.<lambda> at 0x14777fdc0> if I do it that way.  

Is there any advice on how to introduce customize loss function in QAOA? Thanks

**Comment by scarrazza** [comment]
@shangtai I believe the safest approach is to pass a custom loss function as argument and use it directly with the minimizer. If `loss=None` you just use the default function you have quoted:
```python
if loss is None:
    def _loss(params, ...) # default loss function
else:
    # check loss arguments, if pass then
    _loss =  loss
```
